File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/proxy/solution/ProxyFoo.java
// ProxyFoo.java

public class ProxyFoo extends Foo
{
	public ProxyFoo(RealFoo aRealFoo)
	{
		setRealFoo(aRealFoo);
	}

	public void bar()
	{
		System.out.println("before:");

		myRealFoo.bar();

		System.out.println("after:");
	}

	private void setRealFoo(RealFoo aRealFoo)
	{
		if (aRealFoo == null)
		{
			throw new DpException("void setRealFoo(RealFoo aRealFoo) -- aRealFoo == null.");
		}

		myRealFoo = aRealFoo;
	}

	private RealFoo myRealFoo;
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_cpp_labs_and_solutions/proxy/solution/proxy_foo.cpp
// proxy_foo.cpp

#include <iostream>
#include "dp_exception.h"
#include "proxy_foo.h"
#include "real_foo.h"

proxy_foo::proxy_foo(real_foo * a_real_foo)
{
	set_real_foo(a_real_foo);
}

void proxy_foo::bar()
{
	std::cout << "before:" << std::endl;

	my_real_foo->bar();

	std::cout << "after:" << std::endl;
}

void proxy_foo::set_real_foo(real_foo * a_real_foo)
{
	if (a_real_foo == nullptr)
	{
		throw dp_exception("void proxy_foo::set_real_foo(real_foo * a_real_foo) -- a_real_foo == nullptr.");
	}

	my_real_foo = a_real_foo;
}
